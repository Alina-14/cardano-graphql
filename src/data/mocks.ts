import { Block, Transaction } from '../graphql_types'

// Todo: Use faker to build this dynamically

export const transactions = [{
  id: 'tx2',
  fee: 10,
  inputs: [{
    outpoint: {
      txId: 'tx1',
      outputIndex: 0
    }
  }],
  outputs: [{
    value: 100,
    address: 'pk1'
  }]
}, {
  id: 'tx3',
  fee: 5,
  inputs: [{
    outpoint: {
      txId: 'tx2',
      outputIndex: 0
    }
  }],
  outputs: [{
    value: 50,
    address: 'pk2'
  }]
}] as Transaction[]

export const blocks = [{
  id: 'block1',
  transactions
}] as Block[]
