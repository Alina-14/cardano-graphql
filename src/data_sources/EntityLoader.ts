import * as DataLoader from 'dataloader'
import { Entity } from '../graphql_types'

export function EntityLoader<E extends Entity> (data: E[]) {
  return new DataLoader<E['id'], E>((ids: E['id'][]) => {
    const entities = data.filter(entity => ids.includes(entity.id))
    if (entities.length === 0) return Promise.resolve([null])
    const map = new Map<E['id'], Entity>()
    entities.forEach(entity => map.set(entity['id'], entity))
    return Promise.resolve(ids.map(k => map.get(k) || null))
  })
}
