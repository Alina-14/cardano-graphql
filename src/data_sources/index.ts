export * from './ledger'
export * from './mempool'
