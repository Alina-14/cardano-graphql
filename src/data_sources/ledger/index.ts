export * from './entities'
export * from './Ledger'
