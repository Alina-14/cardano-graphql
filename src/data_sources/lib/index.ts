export * from './BufferTransformer'
export * from './alignDataLoaderValues'
