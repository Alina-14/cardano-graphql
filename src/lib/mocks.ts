import { Block, Epoch, Slot, Transaction } from '../graphql_types'

export const outputs = [{
  value: 100,
  address: 'pk1'
}, {
  value: 200,
  address: 'pk1'
}]

export const transactions = [{
  fee: 10,
  hash: 'someHash',
  id: 'tx1',
  inputs: [{
    txId: 'tx1',
    outputIndex: 0
  }],
  outputs
}, {
  fee: 20,
  hash: 'someHash2',
  id: 'tx2',
  inputs: [{
    txId: 'tx1',
    outputIndex: 0
   }],
  outputs
}] as Transaction[]

export const blocks = [{
  id: 'block1',
  hash: 'someHash',
  merkleRootHash: 'someHash',
  number: 1,
  size: 19,
  slot: slots[0],
  transactions: transactions
}] as Block[]

export const slots = [{
  blocks: blocks,
  fees: 10,
  hash: 'someHash',
  number: 1,
  totalOutput: 1.0,
  transactions: transactions
}] as Slot[]



export const epochs = [{
  number: 1,
  slots: slots
}] as Epoch[]
