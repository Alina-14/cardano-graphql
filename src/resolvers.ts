import { Resolvers } from './graphql_types'

export const resolvers: Resolvers = {
  Query: {
    mempool: async (_root, _args, { dataSources: { mempool } }) => ({
      transactions: await mempool.transactions(),
      transactionCount: await mempool.transactionCount()
    }),
    transaction: (_root, args, { dataSources: { transactions } }) => transactions.findById(args.id),
    transactions: (_root, args, { dataSources: { transactions } }) => transactions.findByIds(args.ids)
  }
}
