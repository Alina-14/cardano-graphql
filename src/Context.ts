import { Ledger } from './data_sources/ledger'

export type Context = {
  dataSources: {
    ledger: Ledger
  }
}
