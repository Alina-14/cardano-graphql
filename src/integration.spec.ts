import { expect } from 'chai'
import * as path from 'path'
import * as fs from 'fs'
import { createTestClient, ApolloServerTestClient } from 'apollo-server-testing'
import { ApolloServerBase } from 'apollo-server-core'
import { Mempool, Ledger } from './data_sources'
import { blocks, transactions } from './lib/mocks'
import { resolvers } from './resolvers'
import { Context } from './Context'

describe('Integration', () => {
  let server: ApolloServerBase
  let client: ApolloServerTestClient

  beforeEach(() => {
    server = new ApolloServerBase({
      dataSources (): Context['dataSources'] {
        return {
          ledger: new Ledger({ blocks, transactions }),
          mempool: new Mempool({ transactions })
        }
      },
      introspection: true,
      resolvers,
      typeDefs: fs.readFileSync(path.join(__dirname, 'schema.graphql'), 'UTF8')
    })
    client = createTestClient(server)
  })

  describe('Ledger', () => {
    describe('Transactions', () => {
      it('Returns a transaction by ID, or null data value if not found', async () => {
        const result = await client.query({
          query: `query { transaction(id: "tx3") { id fee }}`
        })
        const notFoundResult = await client.query({
          query: `query { transaction(id: "tx?") { id fee }}`
        })
        expect(result.data.transaction).to.eql({ id: 'tx3', fee: 5 })
        expect(notFoundResult.data).to.eql(null)
      })
      it('Returns transactions by IDs, returning null data value for no results, or null values in the array if a partial set is found ', async () => {
        const result = await client.query({
          query: `query { transactions(ids: ["tx3", "tx2"]) { id fee } }`
        })
        const notFoundResult = await client.query({
          query: `query { transaction(id: "tx?") { id fee } }`
        })
        const resultWithMissingTxs = await client.query({
          query: `query { transactions(ids: ["tx3", "tx?", "tx2"]) { id fee } }`
        })
        expect(result.data.transactions).to.eql([{ id: 'tx3', fee: 5 }, { id: 'tx2', fee: 10 }])
        expect(notFoundResult.data).to.eql(null)
        expect(resultWithMissingTxs.data.transactions).to.eql([{ id: 'tx3', fee: 5 }, null, { id: 'tx2', fee: 10 }])
      })
    })
  })
})
