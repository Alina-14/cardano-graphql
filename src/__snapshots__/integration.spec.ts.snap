// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration Ledger transactions Returns transactions by IDs, or an array with null values if not found  1`] = `
Object {
  "data": Object {
    "transactions": Array [
      Object {
        "fee": 0,
        "id": "6ac19b8efd7114eea29080064b1ec6b5a10346a6212ee338d46f98b733851e3b",
        "inputs": Array [
          Object {
            "address": "3e44c7393b8c1fa97516c4aaa7667a6ad4b8ee79e0bd5a2e005fcc27",
            "sourceTxId": "3ec59e9b74e297f4a60ea026baa225ce4ae8fde2b017ad1eb2b691acc1d0a843",
            "sourceTxIndex": 0,
            "value": "1084654000000",
          },
        ],
        "outputs": Array [
          Object {
            "address": "4b4853a4c6e9fe3f56d816445a304fb6671423420343f7b62e5f484e",
            "value": "1084654000000",
          },
        ],
      },
      Object {
        "fee": 0,
        "id": "3ec59e9b74e297f4a60ea026baa225ce4ae8fde2b017ad1eb2b691acc1d0a843",
        "inputs": Array [],
        "outputs": Array [
          Object {
            "address": "3e44c7393b8c1fa97516c4aaa7667a6ad4b8ee79e0bd5a2e005fcc27",
            "value": "1084654000000",
          },
        ],
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Integration Ledger transactions Returns transactions by IDs, or an array with null values if not found  2`] = `
Object {
  "data": Object {
    "transactions": Array [],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Integration Ledger transactions Returns transactions by IDs, or an array with null values if not found  3`] = `
Object {
  "data": Object {
    "transactions": Array [
      Object {
        "fee": 0,
        "id": "3ec59e9b74e297f4a60ea026baa225ce4ae8fde2b017ad1eb2b691acc1d0a843",
        "inputs": Array [],
        "outputs": Array [
          Object {
            "address": "3e44c7393b8c1fa97516c4aaa7667a6ad4b8ee79e0bd5a2e005fcc27",
            "value": "1084654000000",
          },
        ],
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
